source/common/stats/thread_local_store.cc

```diff
@@ -198,22 +202,40 @@ void ThreadLocalStoreImpl::releaseScopeCrossThread(ScopeImpl* scope) {
   // reference elements of SharedStatNameStorageSet. So simply swap out the set
   // contents into a local that we can hold onto until the TLS cache is cleared
   // of all references.
+  //
+  // We use a raw pointer here as it's easier to capture it in in the lambda.
   auto rejected_stats = new StatNameStorageSet;
   rejected_stats->swap(scope->central_cache_.rejected_stats_);
-  const uint64_t scope_id = scope->scope_id_;
-  auto clean_central_cache = [this, rejected_stats]() {
-    rejected_stats->free(symbolTable());
-    delete rejected_stats;
-  };
 
   // This can happen from any thread. We post() back to the main thread which will initiate the
   // cache flush operation.
   if (!shutting_down_ && main_thread_dispatcher_) {
+    const uint64_t scope_id = scope->scope_id_;
+
+    // We must delay the cleanup of the rejected stats storage until all the
+    // thread-local caches are cleared. This happens by post(), and it's
+    // possible that post() will not run, such as when an exception is thrown
+    // during startup. To avoid leaking memory and thus failing tests when
+    // this occurs, we hold the rejected stats in 'purgatory', so they can
+    // be cleared out in the ThreadLocalStoreImpl destructor. We'd prefer
+    // to release the memory immediately, however, in which case we remove
+    // the rejected stats set from purgatory.
+    rejected_stats_purgatory_.insert(rejected_stats);
+    auto clean_central_cache = [this, rejected_stats]() {
+      {
+        Thread::LockGuard lock(lock_);
+        rejected_stats_purgatory_.erase(rejected_stats);
+      }
+      rejected_stats->free(symbolTable());
```

Comment:
Why can't you do a `deferredDelete()` on the current thread/dispatcher to do this cleanup?

Reply:
Good question. 2 reasons:

1. I had never heard of the deferredDelete() until your comment here :)
2. Having looked at the impl I think switching from post() to deferredDelete() will not fix the problem that this PR solves. In the test that I added here (and the ones we have not in OSS) there is a throw which prevents normal shutdown flow, and event-loops do not get run; just destructors.

Reply:
I think it might work, since the unique_ptr donated at https://github.com/envoyproxy/envoy/blob/f35eea732cf64b19c7684c673580413840aad0ba/source/common/event/dispatcher_impl.cc#L143 should get cleaned up in the destructor.

Reply:
Chatted IRL, this is a more sophisticated use case than what `deferredDelete()` can easily support.